src/backend/routers/agent.py

```diff
@@ -177,3 +218,166 @@ async def delete_agent(
+
+    Raises:
+        HTTPException: If the agent tool metadata retrieval fails.
+    """
+    try:
+        return agent_tool_metadata_crud.get_all_agent_tool_metadata_by_agent_id(
+            session, agent_id
+        )
+    except Exception as e:
+        raise HTTPException(status_code=500, detail=str(e))
+
+
+@router.post(
+    "/{agent_id}/tool-metadata",
+    response_model=AgentToolMetadata,
+)
+def create_agent_tool_metadata(
+    session: DBSessionDep,
+    agent_id: str,
+    agent_tool_metadata: CreateAgentToolMetadata,
+    request: Request,
+) -> AgentToolMetadata:
+    """
+    Create an agent tool metadata.
+
+    Args:
+        session (DBSessionDep): Database session.
+        agent_id (str): Agent ID.
+        agent_tool_metadata (CreateAgentToolMetadata): Agent tool metadata data.
+        request (Request): Request object.
+
+    Returns:
+        AgentToolMetadata: Created agent tool metadata.
+
+    Raises:
+        HTTPException: If the agent tool metadata creation fails.
+    """
+    user_id = get_header_user_id(request)
+
+    agent_tool_metadata_data = AgentToolMetadataModel(
+        user_id=user_id,
+        agent_id=agent_id,
+        tool_name=agent_tool_metadata.tool_name,
+        type=agent_tool_metadata.type,
+        artifacts=agent_tool_metadata.artifacts,
+    )
+
+    request.state.agent_tool_metadata = agent_tool_metadata_data
+    try:
+        created_agent_tool_metadata = (
+            agent_tool_metadata_crud.create_agent_tool_metadata(
+                session, agent_tool_metadata_data
+            )
+        )
+        request.state.agent_tool_metadata = agent_tool_metadata_data
+    except Exception as e:
+        raise HTTPException(status_code=500, detail=str(e))
+
+    return created_agent_tool_metadata
+
+
+@router.put("/{agent_id}/tool-metadata/{agent_tool_metadata_id}")
+async def update_agent_tool_metadata(
+    agent_id: str,
+    agent_tool_metadata_id: str,
+    session: DBSessionDep,
+    new_agent_tool_metadata: UpdateAgentToolMetadata,
+    request: Request,
+) -> AgentToolMetadata:
+    """
+    Update an agent tool metadata by ID.
+
+    Args:
+        agent_id (str): Agent ID.
+        agent_tool_metadata_id (str): Agent tool metadata ID.
+        session (DBSessionDep): Database session.
+        new_agent_tool_metadata (UpdateAgentToolMetadata): New agent tool metadata data.
+        request (Request): Request object.
+
+    Returns:
+        AgentToolMetadata: Updated agent tool metadata.
+
+    Raises:
+        HTTPException: If the agent tool metadata with the given ID is not found.
+        HTTPException: If the agent tool metadata update fails.
+    """
+    agent_tool_metadata = agent_tool_metadata_crud.get_agent_tool_metadata_by_id(
+        session, agent_tool_metadata_id
+    )
+    if not agent_tool_metadata:
+        raise HTTPException(
+            status_code=400,
+            detail=f"Agent tool metadata with ID {agent_tool_metadata_id} not found.",
```

Comment:
Def not in the scope of this PR, but eventually we should take some time to standardize our error messages